### PR TITLE
Libboostasio fix

### PR DESCRIPTION
**Pull request by lion10243**
My suggested fix for issue #164 

I've also replaced the macros with short templated functions to avoid the preprocessor code gen.

### CODE REVIEW

File: include/libboostasio.h [hunk 1]
```diff
@@ -27,33 +26,35 @@
 #include <boost/bind.hpp>
 
 
-///////////////////////////////////////////////////////////////////
```
Comment (zerodefect) [inline include/libboostasio.h]: I'd wrap this in an anonymous namespace.

File: include/libboostasio.h [hunk 2]
```diff
@@ -161,7 +160,13 @@ class LibBoostAsioHandler : public virtual TcpHandler
...
 #endif
                                             connection,
-                                            fd)));
+                                            fd);
+                _socket.async_read_some(
+                    boost::asio::null_buffers(),
+                    [f = func, strand = _strand](const boost::system::error_code &ec, const std::size_t bytes_transferred) {
+                        strand_sock_dispatch_func(f, strand, ec, bytes_transferred);
+                    }
+                );
```
Comment (zerodefect) [inline include/libboostasio.h]: We could minimise code duplication by adding a member function like the following:

```
std::function<void(boost::system::error_code,std::size_t)> get_io_handler(const bool bRead,
																		  TcpConnection *const connection,
																		  const int fd)
{
	auto func = boost::bind(bRead ? &Watcher::read_handler : &Watcher::write_handler,
							this,
							boost::arg<1>(),
							boost::arg<2>(),
// C++17 has 'weak_from_this()' support.
#if __cplusplus >= 201701L
							weak_from_this(),
#else
							shared_from_this(),
#endif
							connection,
							fd);
	return [f = func, strand = _strand](const boost::system::error_code &ec, const std::size_t bytes_transferred)
	{
		strand_sock_dispatch_func(f, strand, ec, bytes_transferred);
	};
}
```

...the you could just call:

`
_socket.async_write_some(boost::asio::null_buffers(), get_io_handler(true, connection, fd));
`

You should then be able to do something very similar for the timer

Note: Not sure if this compiles.

File: include/libboostasio.h [hunk 3]
```diff
@@ -27,33 +26,35 @@
...
-    if (!apStrand)                                                                              \
-    {                                                                                           \
-        fn(boost::system::errc::make_error_code(boost::system::errc::operation_canceled),std::size_t{0}); \
-        return;                                                                                 \
-    }                                                                                           \
-                                                                                                \
-    apStrand->dispatch(boost::bind(fn,ec,bytes_transferred));                                   \
+template<class Func>
+void strand_sock_dispatch_func(
+    Func fn,
```
Comment (zerodefect) [inline include/libboostasio.h]: I'd make this `Func &&fn` and make use of move semantics.

**Comment by zerodefect** [comment]
I'd suggest a few changes. I agree with the spirit of what you're trying to achieve though.



For the record, I'm not the gatekeeper around here, @EmielBruijntjes has the clout.

File: include/libboostasio.h (hunk 1 above)
Comment (lion10243) [reply include/libboostasio.h]: True, that I will add.

File: include/libboostasio.h (hunk 3 above)
Comment (lion10243) [reply include/libboostasio.h]: well spotted, thanks

File: include/libboostasio.h (hunk 2 above)
Comment (lion10243) [reply include/libboostasio.h]: I like the suggestion - will see if it can be integrated.

**Comment by lion10243** [comment]
@zerodefect  the updated version of the code is a bit cleaner and integrates the changes. The ternary bRead ? bind(read_handler,...) : bind(write_handler, ...)  wouldn't compile on either gcc or clang, so I wrapped the binding of read_handler and write_handler into separate, encasulating functions.

**Comment by lion10243** [comment]
@EmielBruijntjes  Hello Emiel, my colleage and I, here in Berlin, have been users of the AMQP-CPP library for the past 1.5 years. With the recent 2.8.0 release we have unfortunately not been able to use the boost::asio handler in libboostasio.h out-of-the-box. Would you be willing to review the pull request, please, for inclusion in a future minor release of the library?

**Pull request merged**